net/initiate.go

```diff
@@ -126,29 +123,35 @@ func (h *Host) handleProtocolStream(stream libp2pnetwork.Stream) {
 		return
 	}
 
-	var s SwapState
-	s, resp, err := h.makerHandler.HandleInitiateMessage(curPeer, im)
-	if err != nil {
-		log.Warnf("failed to handle protocol message: err=%s", err)
+	h.swapMu.Lock()
+	if h.swaps[im.OfferID] != nil {
+		log.Warnf("ignoring attempting initiation of swap %s: %s", im.OfferID, errSwapAlreadyInProgress)
+		h.swapMu.Unlock()
 		_ = stream.Close()
 		return
 	}
 
-	if err := p2pnet.WriteStreamMessage(stream, resp, stream.Conn().RemotePeer()); err != nil {
-		log.Warnf("failed to send response to peer: %s", err)
-		if err = s.Exit(); err != nil {
-			log.Warnf("Swap exit failure: %s", err)
-		}
+	// set the stream here but not the swapState, since we don't have it yet
+	// HandleInitiateMessage requires the network to be aware of the swap's stream,
+	// since it sends the SendKeysMessage response using that stream.
+	h.swaps[im.OfferID] = &swap{
+		stream: stream,
+	}
+	h.swapMu.Unlock()
+
+	s, err := h.makerHandler.HandleInitiateMessage(curPeer, im)
+	if err != nil {
+		log.Warnf("failed to handle protocol message: err=%s", err)
```

Comment:
It would be nice if we used a warning level log if something failed on our end, but a low log level for bad requests, since those are not our problem. We don't have that level of knowledge this high in the stack and that change is probably out of scope for this PR, but it is something we should be thinking about.